sky/server/server.py

```diff
@@ -102,6 +103,73 @@ def _add_timestamp_prefix_for_server_logs() -> None:
 # response will block other requests from being processed.
 
 
+def _basic_auth_401_response(content: str):
+    """Return a 401 response with basic auth realm."""
+    return fastapi.responses.JSONResponse(
+        status_code=401,
+        headers={'WWW-Authenticate': 'Basic realm=\"SkyPilot\"'},
+        content=content)
+
+
+async def _override_user_info_in_request_body(request: fastapi.Request,
```

Comment:
TODO for the future: We should have the username and password to always use the `request.state` instead of having two ways to handle it.